Output filename appears to require an absolute path

## Bug Report

### Steps to Reproduce:

The following works just fine:

```
>java -jar generator.jar generate profile.json "c:\temp\generator\bar.csv"
```

However, if you use a relative path for the output file, it barfs:

```
C:\temp\generator>java -jar generator.jar generate profile.json baz.csv
```
```
Exception in thread "main" picocli.CommandLine$ExecutionException: Error while r
unning command (com.scottlogic.deg.generator.CommandLine.GenerateCommandLine@574
5ca0e): java.lang.NullPointerException
        at picocli.CommandLine.execute(CommandLine.java:1051)
        at picocli.CommandLine.access$900(CommandLine.java:142)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:1246)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:1214)
        at picocli.CommandLine$AbstractParseResultHandler.handleParseResult(Comm
andLine.java:1122)
        at picocli.CommandLine.parseWithHandlers(CommandLine.java:1405)
        at picocli.CommandLine.parseWithHandler(CommandLine.java:1344)
        at com.scottlogic.deg.generator.App.main(App.java:18)
Caused by: java.lang.NullPointerException
        at com.scottlogic.deg.generator.outputs.dataset_writers.CsvDataSetWriter
.openWriter(CsvDataSetWriter.java:20)
        at com.scottlogic.deg.generator.outputs.dataset_writers.CsvDataSetWriter
.openWriter(CsvDataSetWriter.java:13)
        at com.scottlogic.deg.generator.outputs.targets.FileOutputTarget.outputD
ataset(FileOutputTarget.java:31)
        at com.scottlogic.deg.generator.GenerationEngine.generateDataSet(Generat
ionEngine.java:40)
        at com.scottlogic.deg.generator.GenerateExecute.run(GenerateExecute.java
:31)
```

### Expected Result:

Not barfing

### Actual Result:

barfs

--- a/docs/Schema.md
+++ b/docs/Schema.md
@@ -46,7 +46,7 @@
 			"then": {
 				"allOf": [
 					{ "field": "tariff", "is": "null" },
-					{ "field": "time", "is": "after", "value": "2014-01-01" }
+					{ "field": "time", "is": "after", "value": { "date": "2014-01-01" } }
 				]
 			},
 			"else": { "not": { "field": "tariff", "is": "null" } }

--- a/generator/src/main/java/com/scottlogic/deg/generator/inputs/ProfileReader.java
+++ b/generator/src/main/java/com/scottlogic/deg/generator/inputs/ProfileReader.java
@@ -48,6 +48,10 @@ public class ProfileReader {
                 profileJson,
                 V3ProfileDTO.SchemaVersion);
 
+        if (profileDto.fields == null || profileDto.rules == null) {
+            throw new InvalidProfileException("Profile is invalid, either 'fields' or 'rules' have not been defined.");
+        }
+
         ProfileFields profileFields = new ProfileFields(
             profileDto.fields.stream()
                 .map(fDto -> new Field(fDto.name))

--- a/generator/src/main/java/com/scottlogic/deg/generator/inputs/ProfileReader.java
+++ b/generator/src/main/java/com/scottlogic/deg/generator/inputs/ProfileReader.java
@@ -48,8 +48,11 @@ public class ProfileReader {
                 profileJson,
                 V3ProfileDTO.SchemaVersion);
 
-        if (profileDto.fields == null || profileDto.rules == null) {
-            throw new InvalidProfileException("Profile is invalid, either 'fields' or 'rules' have not been defined.");
+        if (profileDto.fields == null) {
+            throw new InvalidProfileException("Profile is invalid: 'fields' have not been defined.");
+        }
+        if (profileDto.rules == null) {
+            throw new InvalidProfileException("Profile is invalid: 'rules' have not been defined.");
         }
 
         ProfileFields profileFields = new ProfileFields(

--- a/generator/src/main/java/com/scottlogic/deg/generator/outputs/TestCaseGenerationResultWriter.java
+++ b/generator/src/main/java/com/scottlogic/deg/generator/outputs/TestCaseGenerationResultWriter.java
@@ -69,7 +69,8 @@ public class TestCaseGenerationResultWriter {
     }
 
     private void write(ProfileFields fields, Stream<GeneratedObject> dataSet, Path directory, String filenameWithoutExtension) throws IOException {
-        try (Closeable writer = this.datasetWriter.openWriter(directory, filenameWithoutExtension, fields)) {
+        String fileName = this.datasetWriter.getFileName(filenameWithoutExtension);
+        try (Closeable writer = this.datasetWriter.openWriter(directory, fileName, fields)) {
             dataSet.forEach(row -> {
                 try {
                     this.datasetWriter.writeRow(writer, row);

--- a/generator/src/main/java/com/scottlogic/deg/generator/outputs/dataset_writers/CsvDataSetWriter.java
+++ b/generator/src/main/java/com/scottlogic/deg/generator/outputs/dataset_writers/CsvDataSetWriter.java
@@ -16,7 +16,7 @@ public class CsvDataSetWriter implements DataSetWriter<CSVPrinter> {
     private static final CSVFormat writerFormat = CSVFormat.RFC4180;
     private static final CSVFormat csvStringFormatter = writerFormat.withQuoteMode(QuoteMode.ALL);
 
-    public CSVPrinter openWriter(Path directory, String filenameWithoutExtension, ProfileFields profileFields) throws IOException {
+    public CSVPrinter openWriter(Path directory, String fileName, ProfileFields profileFields) throws IOException {
         return writerFormat
             .withEscape('\0') //Dont escape any character, we're formatting strings ourselves
             .withQuoteMode(QuoteMode.NONE)
@@ -24,7 +24,7 @@ public class CsvDataSetWriter implements DataSetWriter<CSVPrinter> {
                 .map(f -> f.name)
                 .toArray(String[]::new))
             .print(
-                directory.resolve(filenameWithoutExtension + ".csv"),
+                directory.resolve(fileName),
                 StandardCharsets.UTF_8);
     }
 
@@ -35,6 +35,11 @@ public class CsvDataSetWriter implements DataSetWriter<CSVPrinter> {
             .collect(Collectors.toList()));
 
         writer.flush();
+    }
+
+    @Override
+    public String getFileName(String fileNameWithoutExtension) {
+        return fileNameWithoutExtension + ".csv";
     }
 
     private static Object extractCellValue(DataBagValue cell){

--- a/generator/src/main/java/com/scottlogic/deg/generator/outputs/dataset_writers/DataSetWriter.java
+++ b/generator/src/main/java/com/scottlogic/deg/generator/outputs/dataset_writers/DataSetWriter.java
@@ -10,8 +10,10 @@ import java.nio.file.Path;
 public interface DataSetWriter<TWriter extends Closeable> {
     TWriter openWriter(
         Path directory,
-        String filenameWithoutExtension,
+        String fileName,
         ProfileFields profileFields) throws IOException;
 
     void writeRow(TWriter writer, GeneratedObject row) throws IOException;
+
+    String getFileName(String fileNameWithoutExtension);
 }

--- a/generator/src/main/java/com/scottlogic/deg/generator/outputs/dataset_writers/JsonDataSetWriter.java
+++ b/generator/src/main/java/com/scottlogic/deg/generator/outputs/dataset_writers/JsonDataSetWriter.java
@@ -22,8 +22,13 @@ public class JsonDataSetWriter implements DataSetWriter<JsonDataSetWriter.JsonWr
     private static final SimpleDateFormat standardDateFormat = new SimpleDateFormat("dd-MM-yyyy hh:mm:ss");
 
     @Override
-    public JsonWriter openWriter(Path directory, String filenameWithoutExtension, ProfileFields profileFields) {
-        return new JsonWriter(directory.resolve(filenameWithoutExtension + ".json"), profileFields);
+    public JsonWriter openWriter(Path directory, String fileName, ProfileFields profileFields) {
+        return new JsonWriter(directory.resolve(fileName), profileFields);
+    }
+
+    @Override
+    public String getFileName(String fileNameWithoutExtension) {
+        return fileNameWithoutExtension + ".json";
     }
 
     @Override

--- a/generator/src/main/java/com/scottlogic/deg/generator/outputs/dataset_writers/MultiDataSetWriter.java
+++ b/generator/src/main/java/com/scottlogic/deg/generator/outputs/dataset_writers/MultiDataSetWriter.java
@@ -7,52 +7,73 @@ import java.io.Closeable;
 import java.io.IOException;
 import java.nio.file.Path;
 import java.util.*;
-import java.util.function.Function;
 import java.util.stream.Collectors;
 
-public class MultiDataSetWriter implements DataSetWriter<MultiDataSetWriter.MultiCloseable> {
-    private final List<DataSetWriter> writers;
+public class MultiDataSetWriter implements DataSetWriter<Closeable> {
+    private final List<StatefulDataSetWriter> writers;
 
     public MultiDataSetWriter(DataSetWriter... writers) {
-        this(Arrays.asList(writers));
-    }
-
-    public MultiDataSetWriter(List<DataSetWriter> writers) {
-        this.writers = writers;
+        this.writers = Arrays.stream(writers).map(StatefulDataSetWriter::new).collect(Collectors.toList());
     }
 
     @Override
-    public MultiCloseable openWriter(Path directory, String filenameWithoutExtension, ProfileFields profileFields) throws IOException {
+    public Closeable openWriter(Path directory, String fileName, ProfileFields profileFields) throws IOException {
         ThrownExceptions exceptions = new ThrownExceptions();
 
-        return exceptions.rethrowIfCaughtOrReturn(new MultiCloseable(
-            this.writers
-                .stream()
-                .collect(Collectors.toMap(
-                    Function.identity(),
-                    writer -> {
-                        try {
-                            return writer.openWriter(directory, filenameWithoutExtension, profileFields);
-                        } catch (IOException e) {
-                            exceptions.add(e);
-                            return new NullClosable();
-                        }
-                    }
-                ))
-        ));
-    }
-
-    @Override
-    public void writeRow(MultiCloseable closeable, GeneratedObject row) throws IOException {
-        ThrownExceptions exceptions = new ThrownExceptions();
-
-        closeable.writers.forEach((writer, writerClosable) -> {
+        this.writers.forEach(writer -> {
             try {
-                writer.writeRow(writerClosable, row);
+                writer.openWriterAndRemember(directory, profileFields);
             } catch (IOException e) {
                 exceptions.add(e);
             }
         });
+
+        exceptions.rethrowIfCaught();
+        return this::closeAllWriters; //to ensure when close() is called then its called on all instances in writers.
+    }
+
+    @Override
+    public void writeRow(Closeable closeable, GeneratedObject row) throws IOException {
+        ThrownExceptions exceptions = new ThrownExceptions();
+
+        this.writers.forEach(writer -> {
+            try {
+                writer.writeRow(row);
+            } catch (IOException e) {
+                exceptions.add(e);
+            }
+        });
+
+        exceptions.rethrowIfCaught();
+    }
+
+    @Override
+    public String getFileName(String fileNameWithoutExtension) {
+        String firstFileName = null;
+        for (StatefulDataSetWriter writerWithFileName : this.writers){
+            String fileName = writerWithFileName.getFileNameAndRemember(fileNameWithoutExtension);
+            if (firstFileName == null && fileName != null){
+                firstFileName = fileName;
+            }
+        }
+
+        return firstFileName;
+    }
+
+    private void closeAllWriters() throws IOException {
+        ThrownExceptions exceptions = new ThrownExceptions();
+
+        for (StatefulDataSetWriter writer : this.writers) {
+            if (writer.closeable == null){
+                continue;
+            }
+
+            try {
+                writer.closeable.close();
+            } catch (IOException e) {
+                exceptions.add(e);
+            }
+        }
 
         exceptions.rethrowIfCaught();
     }
@@ -64,11 +85,6 @@ public class MultiDataSetWriter implements DataSetWriter<MultiDataSetWriter.Mult
             exceptions.add(exception);
         }
 
-        <T> T rethrowIfCaughtOrReturn(T ifNoExceptions) throws IOException {
-            rethrowIfCaught();
-            return ifNoExceptions;
-        }
-
         void rethrowIfCaught() throws IOException {
             if (exceptions.isEmpty()){
                 return;
@@ -78,31 +94,34 @@ public class MultiDataSetWriter implements DataSetWriter<MultiDataSetWriter.Mult
         }
     }
 
-    class NullClosable implements Closeable{
-        @Override
-        public void close() { }
-    }
+    class StatefulDataSetWriter {
+        private final DataSetWriter writer;
+        private String fileName;
+        private Closeable closeable;
 
-    class MultiCloseable implements Closeable {
-        final Map<DataSetWriter, Closeable> writers;
-
-        MultiCloseable(Map<DataSetWriter, Closeable> writers) {
-            this.writers = writers;
+        StatefulDataSetWriter(DataSetWriter writer) {
+            this.writer = writer;
         }
 
-        @Override
-        public void close() throws IOException {
-            ThrownExceptions exceptions = new ThrownExceptions();
-
-            for (Closeable closable : this.writers.values()) {
-                try {
-                    closable.close();
-                } catch (IOException e) {
-                    exceptions.add(e);
-                }
+        void openWriterAndRemember(Path directory, ProfileFields profileFields) throws IOException {
+            if (this.fileName == null){
+                throw new IllegalStateException("Filename has not been determined");
             }
 
-            exceptions.rethrowIfCaught();
+            this.closeable = this.writer.openWriter(directory, this.fileName, profileFields);
+        }
+
+        void writeRow(GeneratedObject row) throws IOException {
+            if (this.closeable == null){
+                throw new IllegalStateException("Writer has not been initialised");
+            }
+
+            this.writer.writeRow(this.closeable, row);
+        }
+
+        String getFileNameAndRemember(String fileNameWithoutExtension) {
+            this.fileName = this.writer.getFileName(fileNameWithoutExtension);
+            return this.fileName;
         }
     }
 }

--- a/generator/src/main/java/com/scottlogic/deg/generator/outputs/dataset_writers/SourceTracingDataSetWriter.java
+++ b/generator/src/main/java/com/scottlogic/deg/generator/outputs/dataset_writers/SourceTracingDataSetWriter.java
@@ -27,9 +27,13 @@ public class SourceTracingDataSetWriter implements DataSetWriter<SourceTracingDa
     }
 
     @Override
-    public JsonArrayOutputWriter openWriter(Path directory, String filenameWithoutExtension, ProfileFields profileFields) throws IOException {
-        File file = directory.resolve(filenameWithoutExtension + "-trace.json").toFile();
-        return new JsonArrayOutputWriter(new PrintWriter(new FileOutputStream(file, false)));
+    public JsonArrayOutputWriter openWriter(Path directory, String fileName, ProfileFields profileFields) throws IOException {
+        return new JsonArrayOutputWriter(new PrintWriter(new FileOutputStream(fileName, false)));
+    }
+
+    @Override
+    public String getFileName(String fileNameWithoutExtension) {
+        return fileNameWithoutExtension + "-trace.json";
     }
 
     @Override

--- a/generator/src/main/java/com/scottlogic/deg/generator/outputs/targets/DirectoryOutputTarget.java
+++ b/generator/src/main/java/com/scottlogic/deg/generator/outputs/targets/DirectoryOutputTarget.java
@@ -31,7 +31,7 @@ public class DirectoryOutputTarget implements OutputTarget {
 
     @Override
     public void outputDataset(Stream<GeneratedObject> generatedObjects, ProfileFields profileFields) throws IOException {
-        try (Closeable writer = this.dataSetWriter.openWriter(this.directoryPath, "output", profileFields)) {
+        try (Closeable writer = this.dataSetWriter.openWriter(this.directoryPath, this.dataSetWriter.getFileName("output"), profileFields)) {
             generatedObjects.forEach(row -> {
                 try {
                     this.dataSetWriter.writeRow(writer, row);

--- a/generator/src/main/java/com/scottlogic/deg/generator/outputs/targets/FileOutputTarget.java
+++ b/generator/src/main/java/com/scottlogic/deg/generator/outputs/targets/FileOutputTarget.java
@@ -10,6 +10,7 @@ import com.scottlogic.deg.generator.outputs.dataset_writers.DataSetWriter;
 import java.io.Closeable;
 import java.io.IOException;
 import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.util.stream.Stream;
 
 public class FileOutputTarget implements OutputTarget {
@@ -29,6 +30,10 @@ public class FileOutputTarget implements OutputTarget {
         throws IOException {
 
         Path directoryPath = this.filePath.getParent();
+        if (directoryPath == null) {
+            directoryPath = Paths.get(System.getProperty("user.dir"));
+        }
+
         String fileNameWithoutExtension = this.filePath.getFileName().toString();
 
         try (Closeable writer = this.dataSetWriter.openWriter(directoryPath, fileNameWithoutExtension, profileFields)) {

--- a/generator/src/test/java/com/scottlogic/deg/generator/smoke_tests/ExampleProfilesTests.java
+++ b/generator/src/test/java/com/scottlogic/deg/generator/smoke_tests/ExampleProfilesTests.java
@@ -124,6 +124,11 @@ class ExampleProfilesTests {
         @Override
         public void writeRow(Closeable closeable, GeneratedObject row) {
         }
+
+        @Override
+        public String getFileName(String fileNameWithoutExtension) {
+            return null;
+        }
     }
 
     @FunctionalInterface

